examples/provider_upgrade_test.go

```diff
@@ -0,0 +1,322 @@
+		ExportStateValidator: func(t *testing.T, state []byte) {
+			writeFile(t, info.stateFile, state)
+			t.Logf("wrote %s", info.stateFile)
+		},
+	})
+	integration.ProgramTest(t, &test)
+	return
+}
+
+func TestProviderUpgrade(t *testing.T) {
+	if testing.Short() {
+		t.Skipf("Skipping in -short mode")
+	}
+	if accept {
+		t.Skipf("Skipping because %s env var is set", acceptEnvVar)
+	}
+	info := newProviderUpgradeInfo(t)
+	t.Logf("Baseline provider version: %s", info.baselineProviderVersion)
+
+	env := []string{}
+	opts := integration.ProgramTestOptions{
+		Dir: info.testCaseDir,
+	}
+
+	ambientProvider, _ := exec.LookPath(providerBinary)
+	require.NotEmptyf(t, ambientProvider, "expected to find a release candidate provider "+
+		"binary in PATH, try to call `make provider` and `export PATH=$PWD/bin:$PATH`")
+
+	previewCounter := 0
+	var pt *integration.ProgramTester
+	test := opts.With(integration.ProgramTestOptions{
+		Env: env,
+		PrePulumiCommand: func(verb string) (func(err error) error, error) {
+			if verb != "preview" {
+				return nil, nil
+			}
+			previewCounter++
+			if previewCounter != 2 {
+				return nil, nil
+			}
+			t.Logf("Importing pre-recorded stateFile")
+			importState(t, pt, info.stateFile)
+			return nil, nil
+		},
+		SkipRefresh:      true,
+		SkipExportImport: true,
+
+		// TODO this test is not quite right. There's two designs for an upgrade test, a
+		// preview-only one design and an full update design. This is currently neither.
+		//
+		// A preview-only design would import V5 state and run pulumi preview on V6 to
+		// assert empty preview. To make this code behave like this we need to run
+		// SkipUpdate: true but this currently messes up with stack name renaming pass that
+		//
+		// SkipUpdate:       true,
+		//
+		// A full update design would provision actual cloud infra on V5 and do a full
+		// pulumi-up on V6. This can be useful if for example update plans are permitted
+		// (only replaces are rejected) but we need to verify that the actual update plans
+		// execute successfully or if the provider update does not touch the infra but
+		// updates the state file and we need to test that.
+		//
+		// What this code does: provision infra on V6, then import V5 state, then do pulumi
+		// up. This mostly works as a preview test, but does unnecessary "up" work.
+	})
+	pt = integration.ProgramTestManualLifeCycle(t, &test)
+	err := pt.TestLifeCycleInitAndDestroy()
+	require.NoError(t, err)
+}
+
+type providerUpgradeInfo struct {
+	testCaseDir             string
+	recordingDir            string
+	baselineProviderVersion string
+	grpcFile                string
+	stateFile               string
+	opts                    integration.ProgramTestOptions
+}
+
+func newProviderUpgradeInfo(t *testing.T) providerUpgradeInfo {
+	info := providerUpgradeInfo{}
+	info.testCaseDir = filepath.Join("testdata", "resources")
+	pyaml := filepath.Join(info.testCaseDir, "Pulumi.yaml")
+	v := parseProviderVersion(t, pyaml)
```

Comment:
I'd leave it, I'm not sure how to avoid copy-pasting resource configurations, perhaps we can parameterize the stacks somehow or resort to YAML munging. We might also need to break down this single stack into N stacks. I'd like to check in a reasonably working first version before making those extensions though.